3.analyze_data/utils/well_significance_testing.py

```diff
@@ -0,0 +1,206 @@
+
+    _dmso_probs: pandas.Series
+        The down-sampled predicted probilities of DMSO for a treatment type and phenotype.
+    _treatment_probs: pandas.Series
+        The predicted probabilities of the treatment
+
+    **_comp_names: Keywork arguments
+        Additional treatment data to include in the results as specified in _treatments
+
+    Returns
+    -------
+    _treatments: Dictionary of Lists
+        The treatment results, which contains keys corresponding to the statistical test, the comparison metric, the p value, and the comparison metric value among other keys specified by comp_names.
+    """
+    for func_name, func_data in _comp_functions.items():
+
+        # Compute the results of the function use prespeficied parameters
+        results = func_data["statistical_test_function"](_dmso_probs, _treatment_probs)
+
+        _treatments["statistical_test"].append(func_name)
+        _treatments["comparison_metric"].append(func_data["comparison_metric"])
+
+        # Store subset of results as predetermined
+        for name, val in results:
+            _treatments[name].append(val)
+
+        # Store the other data associated with the results
+        for name, val in _comp_names.items():
+            _treatments[name].append(val)
+
+    return _treatments
+
+def filter_wells_by_cell_count(_df, _cutoff):
+    """
+    Parameters
+    ----------
+    _df: pandas.Dataframe
+        The dataframe which contains the wells, which will be potentially removed
+
+    _cutoff: Integer
+        The number of cells a well must have to remain in the dataframe
+
+    Returns
+    -------
+    _df: pandas.Dataframe
+        The filtered dataframe with only wells that have more cells than the cutoff
+    """
+
+    # Find the number of cells in each well
+    well_counts = _df["Metadata_Well"].value_counts()
+
+    # Find the wells that contain more cells than the cutoff
+    wells_kept = well_counts[well_counts > _cutoff].index
+
+    # Keep only the wells that contain more cells than the cutoff
+    _df = _df.loc[_df["Metadata_Well"].isin(wells_kept)]
+
+    return _df
+
+def strat_samp_wells(_welldf):
+    """
+    Parameters
+    ----------
+    _welldf: pandas.Dataframe
+        The dataframe which contains the wells, which will be stratify sampled by well
+
+    Returns
+    -------
+    The sampled cells stratified by well
+    """
+
+    # Find the number of cells in each well
+    total_cell_count = _welldf.shape[0]
+
+    def samp_well(_well_samp):
+        well_frac = _well_samp.shape[0] / total_cell_count
+        return _well_samp.sample(frac=well_frac, random_state=0)
+
+    return _welldf.groupby('Metadata_Well', group_keys=False).apply(samp_well)
+
+def get_treatment_comparison(_comp_functions, _treatment_paths, _probadf, _barcode_platemapdf, _control_cutoff = 50, _treat_cutoff = 50):
```

Comment:
Massive function, but is compact and straightforward. Good job. Great inline comments throughout too.